Replace mavserver with the mavlink_vehicles submodule

guiccbr opened the pull request:
The mavlink_vehicles library is a more encapsulated and more generic version of the mavserver. This pull request has been created to centralize the review of the modifications that are needed for gzsitl to be compatible with the mavlink_vehicles library, however, it should not be merged until mavlink_vehicles is made fully available (for now it has restrict access under git://git-amr-4.devtools.intel.com/vision-mavserver.git).

guiccbr committed:
Look at target before moving

Send the MAV_CMD_CONDITION_YAW command in order to have the vehicle continually
pointing at the direction of movement.

guiccbr committed:
Add ACTIVE_AIRBORNE state

Also move common variables and routines outside of the switch case blocks and
implement ACTIVE_ON_GROUND state.

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Replace gzsitl_plugin.cc defines with const variables

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Replace a numeric const, fix indent and add rad2deg

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Use the new mavlink_vehicles library as submodule

To replace the old non-generic mavserver.

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

dakerfp commented on gzsitl/CMakeLists.txt:
```diff
@@ -1,21 +1,17 @@
-set(CFLAGS "-fstack-protector-all -fPIE -fPIC -O2 -D_FORTIFY_SOURCE=2")
+set(CUSTOM_COMPILE_FLAGS "${CUSTOM_COMPILE_FLAGS} -Wall")
+set(LDFLAGS "${LDFLAGS} -z noexecstack -z relro -z now")
+set(CFLAGS "${CFLAGS} -fstack-protector-all -fPIE -fPIC -O2 -D_FORTIFY_SOURCE=2")
```
fPIC isn't set by CMake when the target is a shared library?

guiccbr replied:
Apparently yes, according to [cmake](https://cmake.org/cmake/help/v3.0/command/add_library.html). However, -fPIC is an intel requirement for both shared and static libraries, and since the cmake could be changed to make the library static in some cases, I think it's safer just to keep the flag explicit.

dakerfp replied:
Why is it a requirement on static case?

guiccbr replied:
That's a good point. I'm not an expert on this matter but from what I've read, in order to successfully generate a position independent executable (PIE), the libraries - both shared and static - that are linked to the executable must be position independent too (compiled with -fPIC). Generating PIEs is a security measure to avoid attacks that depend on knowing beforehand the location where program data will be loaded during execution.

dakerfp commented on gzsitl/CMakeLists.txt:
```diff
@@ -1,21 +1,17 @@
 set(CFLAGS "${CFLAGS} -Wformat -Wformat-security")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${GAZEBO_CXX_FLAGS}")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CFLAGS} ${CUSTOM_COMPILE_FLAGS}")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CFLAGS} ${GAZEBO_CXX_FLAGS}")
```
append cflags and gazebo_cxx_flags in different lines

guiccbr replied:
I'll do it

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -261,10 +287,10 @@ void GZSitlPlugin::OnUpdate()
-            defaults::GZSITL_LOOKAT_ROT_SPEED_DEGPS, -copysign(1, targ_ang), 1,
-            0, 0, 0, COND_YAW_REQUEST_INTERVAL_MS);
+        this->mav->rotate(targ_ang);
+            // MAV_CMD_CONDITION_YAW, fabs(targ_ang),
```
remove comments

guiccbr replied:
nice catch

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -384,7 +365,7 @@ math::Pose GZSitlPlugin::coord_gzlocal_to_mavlocal(math::Pose gzpose)
 
 void GZSitlPlugin::set_global_pos_coord_system(
-    mavlink_global_position_int_t position)
+    global_pos_int position)
```
Do you think that having the int suffix is relevant for the type name?

guiccbr replied:
I think it is, since representing the position in fixed-point int32 is not very intuitive. I myself had to solve a bug when using the api on gzsitl because I got confused with this representation. What we might choose to do is to represent it as doubles instead, but that would be different from the mavlink standard for that type of message, so I'm not very sure what's the best approach.

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
+
+    // Get pointer to the permanent target if exists
+    this->perm_target = model->GetWorld()->GetModel(perm_target_name);
+    if(this->perm_target) {
```
this->target_exists = bool(this->perm_target)

guiccbr replied:
ack

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
+
+    // Set substitute target position in Gazebo if exists
+    this->subs_target = model->GetWorld()->GetModel(subs_target_name);
+    if (subs_target) {
```
this

guiccbr replied:
fixed in fece1325089e8decd867d4f8dc86bf95a56ac12f

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -47,6 +47,56 @@ void GZSitlPlugin::OnUpdate()
+    }
+
+    // Calculate target relative pose if exists
+    math::Pose target_pose_rel;
```
reverse order: rel_target_pose

guiccbr replied:
agree. will do it.

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -15,48 +15,150 @@
+
+    // Get pointer to the permanent target if exists
+    this->perm_target = model->GetWorld()->GetModel(perm_target_name);
+    if (this->perm_target) {
```
this->target_exists = (bool) this->perm_target

guiccbr replied:
Sure, thanks.

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -102,22 +189,18 @@ void GZSitlPlugin::OnUpdate()
             return;
         }
 
-        if (!is_armed) {
```
Why not a switch instead of using negatives ifs?

guiccbr replied:
I'm not sure how I would use a switch here. I need to check sequentially a series of flags to see if a specific command has to be sent/resent.

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -128,56 +211,47 @@ void GZSitlPlugin::OnUpdate()
         // Send Target if exists and if it has been moved
-        if (is_flying() && tpose_new != tpose) {
-            tpose = tpose_new;
+        if (target_pose != target_pose_prev) {
```
if == continue

guiccbr replied:
sure

dakerfp commented on gzsitl/gzsitl_plugin.cc:
```diff
@@ -192,15 +266,33 @@ void GZSitlPlugin::OnUpdate()
+
+        // Otherwise, continue to request the rotation
+        this->mav->rotate(targ_ang);
+            // MAV_CMD_CONDITION_YAW, fabs(targ_ang),
```
Remove comments

guiccbr replied:
ack

guiccbr committed:
Remove unneeded folder

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Fix cmake style

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Remove left-out comments, use pragma once and fix indentation

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Rename target_pose_rel, break if targ==prev_targ

Also replace returns inside switch-cases with breaks.

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Use shorthand operators

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr commented:
@dakerfp, @rchiossi, I thought github would be smart enough to get rid of the commits that had already been merged into master but that's was not the case. Just a heads-up: only commits from Jul, 25 onwards are new.

guiccbr committed:
Rotation and special wp send implemented

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Fix ENU-NED convertion and replace started() with is_ready()

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
Update mavlink_vehicles submodule

Temporarily point to the Pull Request #3 (Implment Rotate, Mission WP and
Detour WP).

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr commented:
See #25
